### PR TITLE
Ensuring that ReactInstance Initialize returns only after initialization of modules is complete

### PR DESCRIPTION
**Pull request by psivaram**
Change Summary:
Changed ReactInstance.Initialize to ReactInstance.InitializeAsync which awaits for Initialization of modules
Swap the call order of ReactInstance.Initialize and ReactInstance.InitializeBridgeAsync
Updated the test code to reflect the new signatures

Problem this solves:
When the modules are initialized, they may start calling JSModules right away (potentially in another thread). So we need to ensure that the bridge is initialized before the modules are initialized.

Further, once ReactInstance.Initialize returns, the caller, naturally, assumes that the ReactInstance is initialized and so is free to start using RNModules, but the RNModules may still be executing their Initialization code on the RNModule thread. We now ensure that ReactInstance.InitializeAsync returns only after all the RNModule's Initialize methods are executed.

### CODE REVIEW

File: ReactWindows/ReactNative.Tests/Bridge/ReactInstanceTests.cs [hunk 1]
```diff
@@ -81,14 +81,14 @@ public class ReactInstanceTests
 
             var instance = await DispatcherHelpers.CallOnDispatcherAsync(() => builder.Build());
             reactContext.InitializeWithInstance(instance);
-            await DispatcherHelpers.RunOnDispatcherAsync(() => instance.Initialize());
+            await DispatcherHelpers.RunOnDispatcherAsync(async () => await instance.InitializeAsync());
```
Comment (rozele) [inline ReactWindows/ReactNative.Tests/Bridge/ReactInstanceTests.cs]: I think we might need to change this to CallOnDispatcherAsync so it waits for the completion of the inner task.

Comment (psivaram) [reply ReactWindows/ReactNative.Tests/Bridge/ReactInstanceTests.cs]: Done

File: ReactWindows/ReactNative.Shared/Bridge/NativeModuleRegistry.cs [hunk 2]
```diff
@@ -156,7 +157,7 @@ internal void WriteModuleDescriptions(JsonWriter writer)
         /// Hook to notify modules that the <see cref="IReactInstance"/> has
         /// been initialized.
         /// </summary>
-        internal void NotifyReactInstanceInitialize()
+        internal Task NotifyReactInstanceInitializeAsync()
```
[the remark this answers is not in the document]
Comment (psivaram) [reply ReactWindows/ReactNative.Shared/Bridge/NativeModuleRegistry.cs]: Updated the function to wait for Initialize to complete even when it is executed on ActionQueue.

File: ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs [hunk 3]
```diff
@@ -26,11 +26,6 @@ public interface IReactInstance : IAsyncCancelable
         /// </summary>
         IReactQueueConfiguration QueueConfiguration { get; }
 
-        /// <summary>
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs]: Can you just change the `Initialize` method on this interface to `InitializeAsync` for now? I suspect we'll be able to eliminate this interface all together soon (see #1899).

Comment (psivaram) [reply ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs]: Done

**Comment by psivaram** [comment]
@rozele Can you please take a look at the updated code?

**Pull request merged**